lib/ts/recipe/thirdParty/providers/apple.ts

```diff
@@ -0,0 +1,179 @@
+            },
+            additionalProperties: false,
+        },
+        required: ["clientId", "clientSecret"],
+        additionalProperties: false,
+    },
+};
+
+export default function Apple(config: TypeThirdPartyProviderAppleConfig): TypeProvider {
+    validateTheStructureOfUserInput(
+        config,
+        InputSchemaTypeThirdPartyProviderAppleConfig,
+        "thirdparty recipe, provider apple",
+        Receipe.RECIPE_ID
+    );
+    const id = "apple";
+
+    function getClientSecret(clientId: string, keyId: string, teamId: string, privateKey: string): string {
+        return jwtSign(
+            {
+                iss: teamId,
+                iat: Math.floor(Date.now() / 1000),
+                exp: Math.floor(Date.now() / 1000) + 86400 * 180, // 6 months
+                aud: "https://appleid.apple.com",
+                sub: clientId,
+            },
+            privateKey.replace(/\\n/g, "\n"),
+            { algorithm: "ES256", keyid: keyId }
+        );
+    }
+
+    async function get(redirectURI: string, authCodeFromRequest: string | undefined): Promise<TypeProviderGetResponse> {
+        let accessTokenAPIURL = "https://appleid.apple.com/auth/token";
+        let clientSecret = getClientSecret(
+            config.clientId,
+            config.clientSecret.keyId,
+            config.clientSecret.teamId,
+            config.clientSecret.privateKey
+        );
+        let accessTokenAPIParams: { [key: string]: string } = {
+            client_id: config.clientId,
+            client_secret: clientSecret,
+            redirect_uri: redirectURI,
+            grant_type: "authorization_code",
+        };
+        if (authCodeFromRequest !== undefined) {
+            accessTokenAPIParams.code = authCodeFromRequest;
+        }
+        let authorisationRedirectURL = "https://appleid.apple.com/auth/authorize";
+        let scopes = ["name", "email"];
+        if (config.scope !== undefined) {
+            scopes.push(...config.scope);
+            scopes = Array.from(new Set(scopes));
+        }
+        let additionalParams =
+            config.authorisationRedirect === undefined || config.authorisationRedirect.params === undefined
+                ? {}
+                : config.authorisationRedirect.params;
+        let authorizationRedirectParams: { [key: string]: string } = {
+            scope: scopes.join(" "),
+            response_mode: "form_post",
+            response_type: "code",
+            redirect_uri: redirectURI,
+            client_id: config.clientId,
+            ...additionalParams,
+        };
+
+        async function getProfileInfo(accessTokenAPIResponse: {
+            access_token: string;
+            expires_in: number;
+            token_type: string;
+            refresh_token: string;
+            id_token: string;
+        }) {
+            let payload = jwtDecode(accessTokenAPIResponse.id_token);
+            if (payload === null) {
+                throw new STError(
```

Comment:
Throw normal errors here.. as these are called in the API which wrap the error in STError. (Same applies for all providers)